content/posts/2021-01-05-stack-clash-protection.md

```diff
@@ -0,0 +1,276 @@
+aliases: [/posts/2021-01-05-stack-clash-protection/]
+---
+
+# Context
+
+Stack clash is an attack that dates back to 2017, when the Qualys Research Team
+released an advisory with a [joint blogpost][QualysAdvisory]. It basically
+exploits large stack allocation (greater than `PAGE_SIZE`) that can lead to
+stack read/write *not* triggering the [stack guard page][StackGuardPage] allocated by the Linux
+Kernel.
+
+Shortly after the advisory got released, GCC provided a [couter-measure][GCCStackClashProtection] activated by `-fstack-clash-protection` that
+basically consists in splitting large allocation in chunks of `PAGE_SIZE`,
+with a probe in each chunk to trigger the kernel stack guard page.
+
+This has been a major security difference between GCC and Clang since then. It
+has even been identified as a blocker by Fedora to move from gcc to clang as the
+compiler for some projects that already made the move upstream, leading to [extra
+maintenance][FedoraClangGCC] for packagers.
+
+Support for this flag [landed in Clang in 2020][CLANGStackClashProtection],
+only for X86, SystemZ and PowerPC. Its implementation is a result of a fruitful
+collaboration between LLVM, Firefox and Rust developpers.
+
+
+Rust already had a countermeasure implemnted in the form of a runtime call to
+perform the stack probing. With LLVM catching up, using a more leightweight
+approach got [investigated in Rust][RustStackProbe].
+
+
+# Countermeasure Description
+
+The Clang implementation for X86 is derived from the GCC implementation, with a
+few distinctions. The core ideas are:
+
+1. thanks to X86 calling convention, we get a free probe at each call site,
+   which means that each function starts with a probed stack
+
+2. when probing the stack in the function prologue, we don't probe the tail of
+   the allocation. Stated otherwise, if the stack size is `PAGE_SIZE + PAGE_SIZE/2`,
+   we want to probe only once. This is important to limit the number
+   of probes: if the stack size is lower than `PAGE_SIZE` no probe is needed
+
+3. because a signal can interrupt the execution flow any time, at no point
+   should we have two stack allocations (lower than `PAGE_SIZE`) without a probe
+   in between.
+
+
+The probing strategy for stack allocation varies based on the size of the stack
+allocation. If it's smaller than `PAGE_SIZE`, thanks to (2) no probing is
+needing. If it's below a small multiple of ``PAGE_SIZE``, then the probing loop
+can be unrolled. Otherwise a probing loop alternates stack allocation of
+`PAGE_SIZE` bytes and probe, starting with the allocation thanks to (1).
+
+As side effect of (2) is that when performing a dynamic allocation, we need to
+probe *before* updating the stack, otherwise we got a hole in the protection.
+This probe cannot be done after the stack update, even with an offset, because
+of (3). Otherwise we end up with a bug as this one found in gcc [GCCProbeBug]_
```

Comment:
I'd recommend normalizing "gcc" to "GCC" unless referring to the command line tool.